### PR TITLE
Allow access modifiers to auto properties getters and setters

### PR DESCRIPTION
**Pull request by Tangent-90**
## Description

Implements [this suggestion](https://github.com/fsharp/fslang-suggestions/issues/430)

图片

## Checklist

- [x] [RFC added](https://github.com/fsharp/fslang-design/pull/764)
- [x] Put it under preview flag
- [x] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/releae-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

**Comment by Tangent-90** [comment]
Where should I add tests to?

File: src/Compiler/SyntaxTree/SyntaxTree.fsi [hunk 1]
```diff
@@ -242,6 +242,52 @@ type SynAccess =
...
+type GetSetKeywords =
+    | Get of range
+    | Set of range
+    | GetSet of get: range * getterAccess: SynAccess option * set: range * setterAccess: SynAccess option
+
+    member Range: range
+
+/// Represents additional information for SynMemberDefn.AutoProperty
+[<NoEquality; NoComparison>]
+type SynMemberDefnAutoPropertyTrivia =
```
Comment (nojaf) [inline src/Compiler/SyntaxTree/SyntaxTree.fsi:256]: Why did you move this? This feels very off.

Comment (Tangent-90) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi:256]: Because I moved `GetSetKeywords` to be after of `SynAccess`

Comment (nojaf) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi:256]: I do not agree here. The additional SynAccess is not trivia.

Comment (Tangent-90) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi:256]: The additional `SynAccess` was firstly saved into `GetSetKeywords` by parser, then be passed to `mkAutoPropDefn` , so `GetSetKeywords` and types related to it need to be moved after `SynAccess`.

Or could you provide a way to avoid this?

Comment (nojaf) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi:256]: You can add it as an extra argument to `mkAutoPropDefn`. This means you'll have to move some things around in the parser, but the goal is to leave the trivia alone because what you're adding isn't considered trivia. Think of trivia as stuff that's nice for tools to use but not something the compiler needs to have to type-check.

Comment (Tangent-90) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi:256]: will rework this

**Comment by vzarytovskii** [comment]
> Where should I add tests to? 

They should go to ComponentTests probably. Also, will need a small rfc since it changes spec.

**Comment by edgarfgp** [comment]
> Where should I add tests to?

I think this would be a good place https://github.com/dotnet/fsharp/tree/main/tests/FSharp.Compiler.ComponentTests/Conformance/BasicGrammarElements/MemberDefinitions

**Comment by Tangent-90** [comment]
Should this also be allowed for abstract properties? Such as:
```fsharp
type IA =
  abstract B: int with get, internal set
```

**Comment by Tangent-90** [comment]
> Should this also be allowed for abstract properties? Such as:
> ```fsharp
> type IA =
>   abstract B: int with get, internal set
> ```

It seems that abstract members doesn't allow access modifiers

File: src/Compiler/SyntaxTree/SyntaxTree.fsi (hunk 1 above)
Comment (Tangent-90) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi]: completed

Comment (nojaf) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi]: Thank you 💪

**Comment by Tangent-90** [comment]
By the way, access modifiers before getter and setter will be ignored and produce a warning in signature files

**Comment by Tangent-90** [comment]
will anyone make reviews on this pr?

**Comment by vzarytovskii** [comment]
> will anyone make reviews on this pr?

Yes, we will, we're a bit short-handed now, a bunch of folks are off.

**Review by vzarytovskii — APPROVED** [review]

**Comment by edgarfgp** [comment]
> By the way, access modifiers before getter and setter will be ignored and produce a warning in signature files

Would this mean that a `private val member with private getter and public setter` would be usable from the outside ? 

```fsharp
type A() =
    member val private X = 0 with public get, private set

let a = A()
a.X //  ? 
```

If that is the case we should a e compiler error saying that there are conflicting access modifiers. cc

**Comment by Tangent-90** [comment]
I found that in ILSpy, `member val private A = 0 with get, set` and `member val A = 0 with internal get, private set` both produce a internal property. Is it correct behavior? 
```csharp
	internal int A
	{
		[CompilerGenerated]
		[DebuggerNonUserCode]
		get
		{
			return A@;
		}
		[CompilerGenerated]
		[DebuggerNonUserCode]
		set
		{
			A@ = value;
		}
	}
```

**Comment by vzarytovskii** [comment]
> I found that in ILSpy, `member val private A = 0 with get, set` and `member val A = 0 with internal get, private set` both produce a internal property. Is it correct behavior?
> 
> ```cs
> 	internal int A
> 	{
> 		[CompilerGenerated]
> 		[DebuggerNonUserCode]
> 		get
> 		{
> 			return A@;
> 		}
> 		[CompilerGenerated]
> 		[DebuggerNonUserCode]
> 		set
> 		{
> 			A@ = value;
> 		}
> 	}
> ```

Yes, current compiler always emits everything (private and internal) as internal.

**Comment by vzarytovskii** [comment]
@Tangent-90 ugh, it seems you're facing some issues with tests post merge, sorry about that. Let me know if I can help resolving those.

**Comment by Tangent-90** [comment]
> @Tangent-90 ugh, it seems you're facing some issues with tests post merge, sorry about that. Let me know if I can help resolving those.

Just some boring error messages position mismatched in tests...

**Comment by Tangent-90** [comment]
@vzarytovskii Still has a test failed.
```
Failed FSharpChecker.TransparentCompiler.File is not checked twice [291 ms] 

Error Message: 

Assert.Equal() Failure 

Expected: FSharpList<JobEvent> [Weakened, Requested, Started, Finished] 

Actual: FSharpList<JobEvent> [Weakened, Requested, Started] 

Stack Trace: 

at FSharpChecker.TransparentCompiler.File is not checked twice() in /home/vsts/work/1/s/tests/FSharp.Compiler.ComponentTests/FSharpChecker/TransparentCompiler.fs:line 242 

at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor) 

at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr) 
```

**Review by vzarytovskii — APPROVED** [review]
Good stuff as always, thanks!

**Review by edgarfgp — APPROVED** [review]

File: src/Compiler/FSComp.txt [hunk 2]
```diff
@@ -1745,4 +1745,6 @@ featureReuseSameFieldsInStructUnions,"Share underlying fields in a [<Struct>] di
 3862,parsStaticMemberImcompleteSyntax,"Incomplete declaration of a static construct. Use 'static let','static do','static member' or 'static val' for declaration."
 3863,parsExpectingField,"Expecting record field"
 3864,tooManyMethodsInDotNetTypeWritingAssembly,"The type '%s' has too many methods. Found: '%d', maximum: '%d'"
-3865,parsOnlySimplePatternsAreAllowedInConstructors,"Only simple patterns are allowed in primary constructors"
\ No newline at end of file
+3865,parsOnlySimplePatternsAreAllowedInConstructors,"Only simple patterns are allowed in primary constructors"
+featureAllowAccessModifiersToAutoPropertiesGettersAndSetters,"Allow access modifiers to auto properties getters and setters"
+3866,parsAccessModifiersBeforeGettersAndSettersNotAllowedInSigFile,"The modifier will be ignored because accessible modifiers before getters and setters are not allowed in signature file."
```
Comment (psfinaki) [inline src/Compiler/FSComp.txt:1750]: I think it should rather be "access modifiers" or "accessibility modifiers". 
I know it's nitpicking, but this is user-facing, so probably worth using the correct term.

Comment (Tangent-90) [reply src/Compiler/FSComp.txt]: It was unified to "access modifiers"

Comment (psfinaki) [reply src/Compiler/FSComp.txt]: Thanks!

**Comment by auduchinok** [comment]
> By the way, access modifiers before getter and setter will be ignored and produce a warning in signature files

Why? Would it be possible to support signatures too?

> member val X = 0 with public get, private set in signature files should be written as following, and can be usable from outside of A

Can we fix this and allow the same syntax, please?

@vzarytovskii I think we should not release the feature without a proper signature support.

**Comment by Tangent-90** [comment]
> > By the way, access modifiers before getter and setter will be ignored and produce a warning in signature files
> 
> Why? Would it be possible to support signatures too?

That might so much thing to change... It also relates to the generation of signature files.
